In [22]:
import os
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt 
import numpy as np
# Configure Notebook

sns.set_context("notebook") 
import warnings 
warnings.filterwarnings('ignore')

In [23]:
import matplotlib as mpl
mpl.rcParams['figure.dpi'] = 300 # increase resolution

# Reading Bikes Data

First, lets read the bikes data and combine it into a single data frame. The bikes data all share a common prefix, so it should be relatively easy to read it in.

In [24]:
trips_filenames = [filename for filename in os.listdir('project_data') if 'bike_share' in filename]
trips_filenames

['bike_share_2019-1.csv',
 'bike_share_2020-5.csv',
 'bike_share_2020-4.csv',
 'bike_share_2020-10.csv',
 'bike_share_2019-2.csv',
 'bike_share_2020-6.csv',
 'bike_share_2020-7.csv',
 'bike_share_2019-3.csv',
 'bike_share_2018-11.csv',
 'bike_share_2017-8.csv',
 'bike_share_2019-7.csv',
 'bike_share_2020-3.csv',
 'bike_share_2020-2.csv',
 'bike_share_2019-6.csv',
 'bike_share_2017-9.csv',
 'bike_share_2018-10.csv',
 'bike_share_2018-12.csv',
 'bike_share_2019-4.csv',
 'bike_share_2020-1.csv',
 'bike_share_2019-5.csv',
 'bike_share_2018-2.csv',
 'bike_share_2018-3.csv',
 'bike_share_2018-1.csv',
 'bike_share_2018-4.csv',
 'bike_share_2018-5.csv',
 'bike_share_2018-7.csv',
 'bike_share_2018-6.csv',
 'bike_share_2017-10.csv',
 'bike_share_2017-11.csv',
 'bike_share_2018-8.csv',
 'bike_share_2017-12.csv',
 'bike_share_2018-9.csv',
 'bike_share_2019-12.csv',
 'bike_share_2019-10.csv',
 'bike_share_2019-11.csv',
 'bike_share_2017-7.csv',
 'bike_share_2019-8.csv',
 'bike_share_2019-9.csv',
 '

The data comes in 2 separate sets of column formats. One with 9 columns, and one with 11 columns. The 9 column dataset (which we'll name `trips_9`) spans 2017-2018, while the 11 column dataset spans 2019-2020 (which we'll name `trips_11`. The 2 sets of formats also have different date formats. 


We'll need to separately parse it. Below will append the file to 2 different lists depending on the number of columns they have.

In [25]:
trips_11col = []
trips_9col = []

for name in trips_filenames:
    df = pd.read_csv('project_data/' + name)
    
    if len(df.columns) == 11:
        
        trips_11col.append(df)
        
    else:
        
        trips_9col.append(df)

In [26]:
trips_11 = pd.concat(trips_11col)
trips_9 = pd.concat(trips_9col)

In [27]:
trips_9

trip_id  trip_duration_seconds  from_station_id  \
0      4357633                    167           7089.0   
1      4357634                    461           7260.0   
2      4357635                    604           7281.0   
3      4357637                   1133           7047.0   
4      4357638                    753           7032.0   
...        ...                    ...              ...   
48958   868761                    577           7021.0   
48959   868762                    292           7069.0   
48960   868763                    709           7010.0   
48961   868764                    699           7058.0   
48962   868765                    572           7047.0   

              trip_start_time                  from_station_name  \
0      2018-11-01 05:01 (UTC)               Church St  / Wood St   
1      2018-11-01 05:03 (UTC)        Spadina Ave / Adelaide St W   
2      2018-11-01 05:03 (UTC)  Charles St W / Balmuto St - SMART   
3      2018-11-01 05:06 (UTC)      University Ave / Gerrard St W   
4      2018-11-01 05:07 (UTC)          Augusta Ave / Dundas St W   
...                       ...                                ...   
48958  2017-04-01 04:42 (UTC)                 Bay St / Albert St   
48959  2017-04-01 04:50 (UTC)           Queen St W / Spadina Ave   
48960  2017-04-01 04:51 (UTC)            King St W / Spadina Ave   
48961  2017-04-01 04:53 (UTC)                  Huron/ Harbord St   
48962  2017-04-01 04:54 (UTC)      University Ave / Gerrard St W   

               trip_stop_time  to_station_id  \
0      2018-11-01 05:04 (UTC)         7077.0   
1      2018-11-01 05:11 (UTC)         7333.0   
2      2018-11-01 05:13 (UTC)         7002.0   
3      2018-11-01 05:25 (UTC)         7140.0   
4      2018-11-01 05:20 (UTC)         7210.0   
...                       ...            ...   
48958  2017-04-01 04:52 (UTC)         7024.0   
48959  2017-04-01 04:55 (UTC)         7037.0   
48960  2017-04-01 05:02 (UTC)         7060.0   
48961  2017-04-01 05:04 (UTC)         7057.0   
48962  2017-04-01 05:03 (UTC)         7044.0   

                                to_station_name      user_type  
0                            College Park South  Annual Member  
1                       King St E / Victoria St  Annual Member  
2                    St. George St / Bloor St W  Annual Member  
3                   Macpherson Ave / Spadina Rd  Annual Member  
4      Mary McCormick Rec Centre (Sheridan Ave)  Annual Member  
...                                         ...            ...  
48958                  Dundonald St / Church St         Member  
48959                 Bathurst St / Dundas St W         Member  
48960                 Princess St / Adelaide St         Member  
48961               Simcoe St / Wellington St W         Member  
48962                  Church St / Alexander St         Member  

[3415324 rows x 9 columns]

In [28]:
trips_11

Trip Id  Subscription Id  Trip  Duration  Start Station Id  \
0       4581278           199751            1547              7021   
1       4581279           294730            1112              7160   
2       4581280           197252             589              7055   
3       4581281           171700             259              7012   
4       4581282           306314             281              7041   
...         ...              ...             ...               ...   
494463  9476679           694267             938              7154   
494464  9476681           628935             592              7012   
494465  9476682           780037             843              7076   
494466  9476683           498701             420              7038   
494467  9476684           780052            1173              7514   

                       Start Time                         Start Station Name  \
0       01/01/2019 00:08:00 (EST)                         Bay St / Albert St   
1       01/01/2019 00:10:00 (EST)                   King St W / Tecumseth St   
2       01/01/2019 00:15:00 (EST)                     Jarvis St / Carlton St   
3       01/01/2019 00:16:00 (EST)    Elizabeth St / Edward St (Bus Terminal)   
4       01/01/2019 00:19:00 (EST)                       Edward St / Yonge St   
...                           ...                                        ...   
494463  31/08/2020 23:59:00 (EST)                    Bathurst Subway Station   
494464  31/08/2020 23:59:00 (EST)    Elizabeth St / Edward St (Bus Terminal)   
494465  31/08/2020 23:59:00 (EST)                    York St / Queens Quay W   
494466  31/08/2020 23:59:00 (EST)                     Dundas St W / Yonge St   
494467  31/08/2020 23:59:00 (EST)  Humber Bay Shores Park / Marine Parade Dr   

        End Station Id                   End Time  \
0               7233.0  01/01/2019 00:33:00 (EST)   
1               7051.0  01/01/2019 00:29:00 (EST)   
2               7013.0  01/01/2019 00:25:00 (EST)   
3               7235.0  01/01/2019 00:20:00 (EST)   
4               7257.0  01/01/2019 00:24:00 (EST)   
...                ...                        ...   
494463          7238.0  01/09/2020 00:15:00 (EST)   
494464          7100.0  01/09/2020 00:09:00 (EST)   
494465          7041.0  01/09/2020 00:13:00 (EST)   
494466          7044.0  01/09/2020 00:06:00 (EST)   
494467          7514.0  01/09/2020 00:19:00 (EST)   

                                 End Station Name  Bike Id      User Type  
0                        King / Cowan Ave - SMART     1296  Annual Member  
1             Wellesley St E / Yonge St (Green P)     2947  Annual Member  
2                        Scott St / The Esplanade     2293  Annual Member  
3         Bay St / College St (West Side) - SMART      283  Annual Member  
4                    Dundas St W / St. Patrick St     1799  Annual Member  
...                                           ...      ...            ...  
494463  Sorauren/Geoffrey (Sorauren Park) - SMART     6836  Annual Member  
494464             Dundas St E / Regent Park Blvd     4125  Annual Member  
494465                       Edward St / Yonge St     5929  Casual Member  
494466                   Church St / Alexander St     4113  Annual Member  
494467  Humber Bay Shores Park / Marine Parade Dr      843  Casual Member  

[5052221 rows x 11 columns]

# Converting Location Data to a GeoDataFrame

First, lets read the csv of the bikeshare stations as a panda.

In [29]:
stations = pd.read_csv('project_data/bikeshare_stations.csv')
stations

Station Id                       Station Name        lat        lon  \
0          7000       Fort York  Blvd / Capreol Ct  43.639832 -79.395954   
1          7001    Lower Jarvis St / The Esplanade  43.647830 -79.370698   
2          7002         St. George St / Bloor St W  43.667333 -79.399429   
3          7003           Madison Ave / Bloor St W  43.667158 -79.402761   
4          7004            University Ave / Elm St  43.656518 -79.389099   
..          ...                                ...        ...        ...   
605        7663             Kilgour Rd / Rumsey Rd  43.718039 -79.371914   
606        7664  Sunnybrook Health Center - L Wing  43.722680 -79.376440   
607        7665  Sunnybrook Health Centre - S Wing  43.720669 -79.377553   
608        7666         Dundas St W / St Helen Ave  43.650422 -79.440765   
609        7667           Spadina Ave / Sussex Ave  43.664733 -79.403028   

     capacity  
0          35  
1          15  
2          19  
3          15  
4          11  
..        ...  
605        17  
606        15  
607        11  
608        26  
609        19  

[610 rows x 5 columns]

Lets convert the stations data to a geodataframe to easily manipulate the geospatial data. The csv has a column for `lat` and `lon` which will make the conversion easy. We'll need to use some additional packages.

In [30]:
import geopandas as gpd
import folium
import pyproj

In [32]:
stations_gdf = gpd.GeoDataFrame(stations, geometry=gpd.points_from_xy(stations.lon, stations.lat))

Lets set the projection to `epsg:4326`, which will make it easy to map.

In [33]:
stations_gdf.set_crs(epsg = '4326', inplace=True)

Station Id                       Station Name        lat        lon  \
0          7000       Fort York  Blvd / Capreol Ct  43.639832 -79.395954   
1          7001    Lower Jarvis St / The Esplanade  43.647830 -79.370698   
2          7002         St. George St / Bloor St W  43.667333 -79.399429   
3          7003           Madison Ave / Bloor St W  43.667158 -79.402761   
4          7004            University Ave / Elm St  43.656518 -79.389099   
..          ...                                ...        ...        ...   
605        7663             Kilgour Rd / Rumsey Rd  43.718039 -79.371914   
606        7664  Sunnybrook Health Center - L Wing  43.722680 -79.376440   
607        7665  Sunnybrook Health Centre - S Wing  43.720669 -79.377553   
608        7666         Dundas St W / St Helen Ave  43.650422 -79.440765   
609        7667           Spadina Ave / Sussex Ave  43.664733 -79.403028   

     capacity                    geometry  
0          35  POINT (-79.39595 43.63983)  
1          15  POINT (-79.37070 43.64783)  
2          19  POINT (-79.39943 43.66733)  
3          15  POINT (-79.40276 43.66716)  
4          11  POINT (-79.38910 43.65652)  
..        ...                         ...  
605        17  POINT (-79.37191 43.71804)  
606        15  POINT (-79.37644 43.72268)  
607        11  POINT (-79.37755 43.72067)  
608        26  POINT (-79.44076 43.65042)  
609        19  POINT (-79.40303 43.66473)  

[610 rows x 6 columns]

We'll create a simple map to QC the locations of the bike share stations.

In [202]:
m = folium.Map(location=[43.653, -79.3832],
    zoom_start=11
)

folium.features.GeoJson(stations_gdf).add_to(m)


In [203]:
m

The station locations look generally fine, and there's no issue with them.

# Wrangling the Trips Dataset

### Merging the Datasets

Lets reconcile the date formats now, starting with converting the 11 column dataset. First, we'll take out the timezone from the start and end times columns.

In [36]:
trips_11['Start Time'] = trips_11['Start Time'].str.replace(' \(EST\)$', '', regex=True)
trips_11['End Time'] = trips_11['End Time'].str.replace(' \(EST\)$', '', regex=True)

Next we need to convert the start and end times to datetime index, and set the time zone (which seems to be EST) to EST.

In [37]:
trips_11['Start Time'] = pd.DatetimeIndex(trips_11['Start Time']).tz_localize(tz = 'EST')
trips_11['End Time'] = pd.DatetimeIndex(trips_11['End Time']).tz_localize(tz = 'EST')

In [38]:
trips_11

Trip Id  Subscription Id  Trip  Duration  Start Station Id  \
0       4581278           199751            1547              7021   
1       4581279           294730            1112              7160   
2       4581280           197252             589              7055   
3       4581281           171700             259              7012   
4       4581282           306314             281              7041   
...         ...              ...             ...               ...   
494463  9476679           694267             938              7154   
494464  9476681           628935             592              7012   
494465  9476682           780037             843              7076   
494466  9476683           498701             420              7038   
494467  9476684           780052            1173              7514   

                      Start Time                         Start Station Name  \
0      2019-01-01 00:08:00-05:00                         Bay St / Albert St   
1      2019-01-01 00:10:00-05:00                   King St W / Tecumseth St   
2      2019-01-01 00:15:00-05:00                     Jarvis St / Carlton St   
3      2019-01-01 00:16:00-05:00    Elizabeth St / Edward St (Bus Terminal)   
4      2019-01-01 00:19:00-05:00                       Edward St / Yonge St   
...                          ...                                        ...   
494463 2020-08-31 23:59:00-05:00                    Bathurst Subway Station   
494464 2020-08-31 23:59:00-05:00    Elizabeth St / Edward St (Bus Terminal)   
494465 2020-08-31 23:59:00-05:00                    York St / Queens Quay W   
494466 2020-08-31 23:59:00-05:00                     Dundas St W / Yonge St   
494467 2020-08-31 23:59:00-05:00  Humber Bay Shores Park / Marine Parade Dr   

        End Station Id                  End Time  \
0               7233.0 2019-01-01 00:33:00-05:00   
1               7051.0 2019-01-01 00:29:00-05:00   
2               7013.0 2019-01-01 00:25:00-05:00   
3               7235.0 2019-01-01 00:20:00-05:00   
4               7257.0 2019-01-01 00:24:00-05:00   
...                ...                       ...   
494463          7238.0 2020-01-09 00:15:00-05:00   
494464          7100.0 2020-01-09 00:09:00-05:00   
494465          7041.0 2020-01-09 00:13:00-05:00   
494466          7044.0 2020-01-09 00:06:00-05:00   
494467          7514.0 2020-01-09 00:19:00-05:00   

                                 End Station Name  Bike Id      User Type  
0                        King / Cowan Ave - SMART     1296  Annual Member  
1             Wellesley St E / Yonge St (Green P)     2947  Annual Member  
2                        Scott St / The Esplanade     2293  Annual Member  
3         Bay St / College St (West Side) - SMART      283  Annual Member  
4                    Dundas St W / St. Patrick St     1799  Annual Member  
...                                           ...      ...            ...  
494463  Sorauren/Geoffrey (Sorauren Park) - SMART     6836  Annual Member  
494464             Dundas St E / Regent Park Blvd     4125  Annual Member  
494465                       Edward St / Yonge St     5929  Casual Member  
494466                   Church St / Alexander St     4113  Annual Member  
494467  Humber Bay Shores Park / Marine Parade Dr      843  Casual Member  

[5052221 rows x 11 columns]

Now let's repeat the process for the 9 column dataset. This time, we're localizing to UTC, since thats what the date was at, and then convert it to EST. 

In [204]:
trips_9['trip_start_time'] = trips_9['trip_start_time'].str.replace(' \(UTC\)$', '', regex=True)
trips_9['trip_stop_time'] = trips_9['trip_stop_time'].str.replace(' \(UTC\)$', '', regex=True)

KeyError: 'trip_start_time'

In [41]:
trips_9['trip_start_time'] = pd.DatetimeIndex(trips_9['trip_start_time']).tz_localize(tz = 'UTC').tz_convert(tz = 'EST')

In [42]:
trips_9['trip_stop_time'] = pd.DatetimeIndex(trips_9['trip_stop_time']).tz_localize(tz = 'UTC').tz_convert(tz = 'EST')

In [43]:
trips_9

trip_id  trip_duration_seconds  from_station_id  \
0      4357633                    167           7089.0   
1      4357634                    461           7260.0   
2      4357635                    604           7281.0   
3      4357637                   1133           7047.0   
4      4357638                    753           7032.0   
...        ...                    ...              ...   
48958   868761                    577           7021.0   
48959   868762                    292           7069.0   
48960   868763                    709           7010.0   
48961   868764                    699           7058.0   
48962   868765                    572           7047.0   

                trip_start_time                  from_station_name  \
0     2018-11-01 00:01:00-05:00               Church St  / Wood St   
1     2018-11-01 00:03:00-05:00        Spadina Ave / Adelaide St W   
2     2018-11-01 00:03:00-05:00  Charles St W / Balmuto St - SMART   
3     2018-11-01 00:06:00-05:00      University Ave / Gerrard St W   
4     2018-11-01 00:07:00-05:00          Augusta Ave / Dundas St W   
...                         ...                                ...   
48958 2017-03-31 23:42:00-05:00                 Bay St / Albert St   
48959 2017-03-31 23:50:00-05:00           Queen St W / Spadina Ave   
48960 2017-03-31 23:51:00-05:00            King St W / Spadina Ave   
48961 2017-03-31 23:53:00-05:00                  Huron/ Harbord St   
48962 2017-03-31 23:54:00-05:00      University Ave / Gerrard St W   

                 trip_stop_time  to_station_id  \
0     2018-11-01 00:04:00-05:00         7077.0   
1     2018-11-01 00:11:00-05:00         7333.0   
2     2018-11-01 00:13:00-05:00         7002.0   
3     2018-11-01 00:25:00-05:00         7140.0   
4     2018-11-01 00:20:00-05:00         7210.0   
...                         ...            ...   
48958 2017-03-31 23:52:00-05:00         7024.0   
48959 2017-03-31 23:55:00-05:00         7037.0   
48960 2017-04-01 00:02:00-05:00         7060.0   
48961 2017-04-01 00:04:00-05:00         7057.0   
48962 2017-04-01 00:03:00-05:00         7044.0   

                                to_station_name      user_type  
0                            College Park South  Annual Member  
1                       King St E / Victoria St  Annual Member  
2                    St. George St / Bloor St W  Annual Member  
3                   Macpherson Ave / Spadina Rd  Annual Member  
4      Mary McCormick Rec Centre (Sheridan Ave)  Annual Member  
...                                         ...            ...  
48958                  Dundonald St / Church St         Member  
48959                 Bathurst St / Dundas St W         Member  
48960                 Princess St / Adelaide St         Member  
48961               Simcoe St / Wellington St W         Member  
48962                  Church St / Alexander St         Member  

[3415324 rows x 9 columns]

To combine the dataframes, we'll need to rename and reorganize the 9 column dataset.

The 9 column dataset doesn't have a column with `Subscription ID` or `Bike ID`, so we'll have to create an empty column to make sure we have 11 columns for both datasets.

In [44]:
trips_9['Subscription Id'] = np.nan
trips_9['Bike Id'] = np.nan

In [46]:
trips_11.head(1)

Trip Id  Subscription Id  Trip  Duration  Start Station Id  \
0  4581278           199751            1547              7021   

                 Start Time  Start Station Name  End Station Id  \
0 2019-01-01 00:08:00-05:00  Bay St / Albert St          7233.0   

                   End Time          End Station Name  Bike Id      User Type  
0 2019-01-01 00:33:00-05:00  King / Cowan Ave - SMART     1296  Annual Member

We also need to rename the columns for `trips_9` to make sure it appends properly to `trips_11`.

In [47]:
trips_9 = trips_9.rename(columns={"trip_id": "Trip Id", "trip_duration_seconds": "Trip Duration", 'from_station_id':'Start Station Id',
                       'trip_start_time': 'Start Time', 'from_station_name':'Start Station Name', 'to_station_id':'End Station Id',
                       'trip_stop_time': 'End Time', 'to_station_name':'End Station Name', 'user_type': 'User Type'})

In [48]:
trips_11 = trips_11.rename(columns={"Trip  Duration": "Trip Duration"})

In [49]:
trips_9 = trips_9[['Trip Id', 'Subscription Id', 'Trip Duration', 'Start Station Id',
       'Start Time', 'Start Station Name', 'End Station Id', 'End Time',
       'End Station Name', 'Bike Id', 'User Type']]

Now we're going to append the 2 dataframes into a master trips dataframes.

In [50]:
trips = trips_9.append(trips_11)
trips

Trip Id  Subscription Id  Trip Duration  Start Station Id  \
0       4357633              NaN            167            7089.0   
1       4357634              NaN            461            7260.0   
2       4357635              NaN            604            7281.0   
3       4357637              NaN           1133            7047.0   
4       4357638              NaN            753            7032.0   
...         ...              ...            ...               ...   
494463  9476679         694267.0            938            7154.0   
494464  9476681         628935.0            592            7012.0   
494465  9476682         780037.0            843            7076.0   
494466  9476683         498701.0            420            7038.0   
494467  9476684         780052.0           1173            7514.0   

                      Start Time                         Start Station Name  \
0      2018-11-01 00:01:00-05:00                       Church St  / Wood St   
1      2018-11-01 00:03:00-05:00                Spadina Ave / Adelaide St W   
2      2018-11-01 00:03:00-05:00          Charles St W / Balmuto St - SMART   
3      2018-11-01 00:06:00-05:00              University Ave / Gerrard St W   
4      2018-11-01 00:07:00-05:00                  Augusta Ave / Dundas St W   
...                          ...                                        ...   
494463 2020-08-31 23:59:00-05:00                    Bathurst Subway Station   
494464 2020-08-31 23:59:00-05:00    Elizabeth St / Edward St (Bus Terminal)   
494465 2020-08-31 23:59:00-05:00                    York St / Queens Quay W   
494466 2020-08-31 23:59:00-05:00                     Dundas St W / Yonge St   
494467 2020-08-31 23:59:00-05:00  Humber Bay Shores Park / Marine Parade Dr   

        End Station Id                  End Time  \
0               7077.0 2018-11-01 00:04:00-05:00   
1               7333.0 2018-11-01 00:11:00-05:00   
2               7002.0 2018-11-01 00:13:00-05:00   
3               7140.0 2018-11-01 00:25:00-05:00   
4               7210.0 2018-11-01 00:20:00-05:00   
...                ...                       ...   
494463          7238.0 2020-01-09 00:15:00-05:00   
494464          7100.0 2020-01-09 00:09:00-05:00   
494465          7041.0 2020-01-09 00:13:00-05:00   
494466          7044.0 2020-01-09 00:06:00-05:00   
494467          7514.0 2020-01-09 00:19:00-05:00   

                                 End Station Name  Bike Id      User Type  
0                              College Park South      NaN  Annual Member  
1                         King St E / Victoria St      NaN  Annual Member  
2                      St. George St / Bloor St W      NaN  Annual Member  
3                     Macpherson Ave / Spadina Rd      NaN  Annual Member  
4        Mary McCormick Rec Centre (Sheridan Ave)      NaN  Annual Member  
...                                           ...      ...            ...  
494463  Sorauren/Geoffrey (Sorauren Park) - SMART   6836.0  Annual Member  
494464             Dundas St E / Regent Park Blvd   4125.0  Annual Member  
494465                       Edward St / Yonge St   5929.0  Casual Member  
494466                   Church St / Alexander St   4113.0  Annual Member  
494467  Humber Bay Shores Park / Marine Parade Dr    843.0  Casual Member  

[8467545 rows x 11 columns]

Now that we have everything in 1 dataset, lets investigate if there are any null values in the dataset.

In [53]:
trips.isnull().sum(axis=0)

Trip Id                     0
Subscription Id       3415324
Trip Duration               0
Start Station Id      1026893
Start Time                  0
Start Station Name        164
End Station Id        1028159
End Time                    1
End Station Name         1409
Bike Id               3415324
User Type                   0
dtype: int64

### Fixing the Station Names

Some Bike Stations have null values, so we'll need to find a way to fill those using the station names. First, lets find how many cases there are with both the station ID and station name missing.

In [54]:
trips[(trips['Start Station Name'].isnull()) & (trips['Start Station Id'].isnull())]

Empty DataFrame
Columns: [Trip Id, Subscription Id, Trip Duration, Start Station Id, Start Time, Start Station Name, End Station Id, End Time, End Station Name, Bike Id, User Type]
Index: []

In [55]:
trips[(trips['End Station Name'].isnull()) & (trips['End Station Id'].isnull())]

Trip Id  Subscription Id  Trip Duration  Start Station Id  \
110062  2302635              NaN              0               NaN   
360     7775292         393055.0            559            7042.0   
3268    7778789         376443.0           1685            7409.0   
3403    7778936         545536.0            964            7424.0   
4456    7780393         448135.0            540            7285.0   
...         ...              ...            ...               ...   
484536  9466001         550766.0              0            7280.0   
484671  9466146         715154.0              0            7270.0   
485080  9466589         778869.0              0            7076.0   
486443  9468063         560315.0              0            7060.0   
490059  9471907         779502.0         240681            7136.0   

                      Start Time                Start Station Name  \
110062 2017-11-29 05:53:00-05:00           Seaton St / Dundas St E   
360    2020-01-05 10:27:00-05:00    Sherbourne St / Wellesley St E   
3268   2020-01-05 19:32:00-05:00           Isabella St / Church St   
3403   2020-01-05 19:48:00-05:00     Merton St / Mount Pleasant Rd   
4456   2020-02-05 09:49:00-05:00  Spadina Ave / Harbord St - SMART   
...                          ...                               ...   
484536 2020-08-31 13:35:00-05:00  Charles St E / Jarvis St - SMART   
484671 2020-08-31 13:43:00-05:00   Church St / Dundas St E - SMART   
485080 2020-08-31 14:11:00-05:00           York St / Queens Quay W   
486443 2020-08-31 15:38:00-05:00       Princess St / Adelaide St E   
490059 2020-08-31 18:23:00-05:00            Queen St W / Close Ave   

        End Station Id                  End Time End Station Name  Bike Id  \
110062             NaN                       NaT              NaN      NaN   
360                NaN 2020-01-05 10:36:00-05:00              NaN     15.0   
3268               NaN 2020-01-05 20:00:00-05:00              NaN   2893.0   
3403               NaN 2020-01-05 20:04:00-05:00              NaN   1665.0   
4456               NaN 2020-02-05 09:58:00-05:00              NaN   1395.0   
...                ...                       ...              ...      ...   
484536             NaN 2020-08-31 13:35:00-05:00              NaN   2769.0   
484671             NaN 2020-08-31 13:43:00-05:00              NaN   5143.0   
485080             NaN 2020-08-31 14:11:00-05:00              NaN   6111.0   
486443             NaN 2020-08-31 15:38:00-05:00              NaN   3227.0   
490059             NaN 2020-03-09 13:15:00-05:00              NaN   5613.0   

            User Type  
110062         Casual  
360     Annual Member  
3268    Annual Member  
3403    Casual Member  
4456    Annual Member  
...               ...  
484536  Annual Member  
484671  Annual Member  
485080  Casual Member  
486443  Annual Member  
490059  Casual Member  

[1267 rows x 11 columns]

These trips represent a negliable number of the overall trips, so we can safely drop them. Without either the station ID or the station name, these trips are functionally useless in our analysis and creates question about data accuracy.

In [56]:
trips = trips.dropna(axis=0, subset=['End Station Name', 'End Station Id'], how = 'all') 

While we're at it, we should drop other rows that are missing data from important columns. Lets drop that 1 row with the missing end time.

In [57]:
trips = trips.dropna(axis=0, subset=['End Time'], how = 'all') 

We'll create `missing_start_station` and `missing_end_station` to create a sub dataset filled with entries with missing station IDs. Then we'll combine them to a single dataframe named `missing_station`.

In [58]:
missing_start_station = trips[trips['Start Station Id'].isnull()][['Start Station Name']].drop_duplicates()
missing_start_station['Data Station Name'] = missing_start_station['Start Station Name']

In [59]:
missing_end_station = trips[trips['End Station Id'].isnull()][['End Station Name']].drop_duplicates()
missing_end_station['Data Station Name'] = missing_end_station['End Station Name']

In [60]:
missing_station = missing_start_station[['Data Station Name']].append(missing_end_station[['Data Station Name']])

To improve the performance of `fuzzywuzzy`, we are going to create a master list of all station names that don't have IDs. This way, we can merge this lookup table with the dataset to fill the IDs, and avoid repeating calculations.

In [61]:
missing_station = missing_station.drop_duplicates()
missing_station = missing_station.reset_index()
missing_station = missing_station.drop(columns = ['index'])
missing_station

Data Station Name
0                   University Ave / Elm St
1                Queen St W / Dovercourt Rd
2                    Roxton Rd / Harbord St
3                 Cherry St / Distillery Ln
4                   Ulster St / Bathurst St
..                                      ...
266        Vanauley St / Queen St W - SMART
267  Summerhill Ave / MacLennan Ave - SMART
268           Bloor St W / Shaw Ave - SMART
269                   Exhibition GO Station
270                Fringe Next Stage - 7219

[271 rows x 1 columns]

In [62]:
import fuzzywuzzy
from fuzzywuzzy import process

For fuzzywuzzy, we are going to use `fuzzywuzzy.process`. This takes in a choices list, a list of what we want to match to.

In [63]:
choices = list(stations['Station Name'])

We'll pre-declare the columns, and fill them with null values.

In [64]:
missing_station['Matched Station Name'] = ''
missing_station['Score'] = np.nan

This will loop through all 271 rows in the `missing_stations` lookup table, and find the best match for the station name (using `extractOne`). Then, it will set the `Matched Station Name` and `Score`.

In [65]:
for index, row in missing_station.iterrows():
    fuzzy_match = process.extractOne(row['Data Station Name'], choices)
    missing_station.at[index, 'Matched Station Name'] = fuzzy_match[0]
    missing_station.at[index, 'Score'] = fuzzy_match[1]

Unfortunately, some results weren't accurate at all. This occurs when the score was less than 90.

In [66]:
missing_station[missing_station['Score'] < 90]

Data Station Name  \
16            Michael Sweet Ave / St. Patrick St   
66           Wellesley St E / Yonge St (Green P)   
74                   Bathurst St / Queens Quay W   
77                   Margueretta St / College St   
91                         Bay St / Davenport Rd   
99                       Simcoe St / Dundas St W   
110                   Bremner Blvd / Spadina Ave   
124  Victoria St / Gould St (Ryerson University)   
138                         Queen St W / Shaw St   
145                     Lansdowne Subway Green P   
149                     Stephenson Ave / Main St   
157                     Bloor St / Brunswick Ave   
166      Woodbine Subway Green P (Cedarvale Ave)   
189    Dockside Dr / Queens Quay E (Sugar Beach)   
190          Ontario Place Blvd / Remembrance Dr   
200                  Queen St E / Larchmount Ave   
207                       Roxton Rd / College St   
209                                 Base Station   
270                     Fringe Next Stage - 7219   

                           Matched Station Name  Score  
16                          King St W / York St   86.0  
66                          King St W / York St   86.0  
74   Sherbourne St / Carlton St (Allan Gardens)   86.0  
77   Sherbourne St / Carlton St (Allan Gardens)   86.0  
91             Wellesley St / Queen's Park Cres   86.0  
99      Elizabeth St / Edward St (Bus Terminal)   86.0  
110   University Ave / Gerrard St W (East Side)   86.0  
124                  St. George St / Bloor St W   86.0  
138             Lower Jarvis St / The Esplanade   86.0  
145  Dundas St W / Roncesvalles Green P - SMART   86.0  
149     Elizabeth St / Edward St (Bus Terminal)   86.0  
157     Elizabeth St / Edward St (Bus Terminal)   86.0  
166                    Madison Ave / Bloor St W   86.0  
189                       King St E / Jarvis St   86.0  
190                      Bremner Blvd / Rees St   86.0  
200  Sherbourne St / Carlton St (Allan Gardens)   86.0  
207     Elizabeth St / Edward St (Bus Terminal)   86.0  
209                      Temperance St. Station   86.0  
270                       Front St E / Scott St   49.0

For these values, we're going have to manually match them since they are hard to match, and produce results that aren't correct. 

For ease of matching, I opened the file of stations in QGIS to locate each unmatched station name, and then find the closest station.

In [67]:
missing_station.iloc[16,1] = 'Simcoe St / Michael Sweet Ave - SMART'
missing_station.iloc[66,1] = 'Yonge St / Dundonald St - SMART'
missing_station.iloc[74,1] = 'Bathurst St/Queens Quay(Billy Bishop Airport)'
missing_station.iloc[77,1] = 'St Clarens Ave / College St'
missing_station.iloc[91,1] = 'Bay St / Scollard St'
missing_station.iloc[99,1] = 'Dundas St W / St. Patrick St'
missing_station.iloc[110,1] = 'Spadina Ave / Fort York Blvd'
missing_station.iloc[124,1] = 'Gould St / Yonge St (Ryerson University)'
missing_station.iloc[138,1] = 'Crawford St / Queen St W'
missing_station.iloc[145,1] = 'Lansdowne Subway Station'
missing_station.iloc[149,1] = 'Danforth Ave / Barrington Ave'
missing_station.iloc[157,1] = 'Borden St / Bloor St W - SMART'
missing_station.iloc[166,1] = 'Woodbine Subway Green P SMART'
missing_station.iloc[189,1] = 'Lower Jarvis / Queens Quay E'
missing_station.iloc[190,1] = 'Ontario Place Blvd / Lake Shore Blvd W (East)'
missing_station.iloc[200,1] = 'Queen St E / Rushbrooke Ave'
missing_station.iloc[207,1] = 'Ossington Ave / College St'
missing_station.iloc[209,1] = None # pretty sure this is a typo
missing_station.iloc[157,1] = 'Bathurst St / Adelaide St W' # according to google, the stage is here

One reason for the mis-match is that some stations have probably moved or changed locations (such as moving to an adjacent intersection), however its still valuable to assign them to the nearest station for the spatial analysis.

In [68]:
missing_station = missing_station.merge(stations[['Station Name', 'Station Id']], left_on = ['Matched Station Name'], right_on = ['Station Name'])

In [69]:
trips = trips.merge(missing_station, left_on=['Start Station Name'], right_on = ['Data Station Name'], how = 'left')

This step will check to see if the initial `Start Station Id` is `Null`, and then fill it with the matched name.

In [70]:
trips['Start Station Id'] = trips['Start Station Id'].where(trips['Start Station Id'].notnull(), trips['Station Id'])

In [71]:
trips = trips[['Trip Id', 'Subscription Id', 'Trip Duration', 'Start Station Id',
       'Start Time', 'Start Station Name', 'End Station Id', 'End Time',
       'End Station Name', 'Bike Id', 'User Type']]

In [72]:
trips.head()

Trip Id  Subscription Id  Trip Duration  Start Station Id  \
0  4357633              NaN            167            7089.0   
1  4357634              NaN            461            7260.0   
2  4357635              NaN            604            7281.0   
3  4357637              NaN           1133            7047.0   
4  4357638              NaN            753            7032.0   

                 Start Time                 Start Station Name  \
0 2018-11-01 00:01:00-05:00               Church St  / Wood St   
1 2018-11-01 00:03:00-05:00        Spadina Ave / Adelaide St W   
2 2018-11-01 00:03:00-05:00  Charles St W / Balmuto St - SMART   
3 2018-11-01 00:06:00-05:00      University Ave / Gerrard St W   
4 2018-11-01 00:07:00-05:00          Augusta Ave / Dundas St W   

   End Station Id                  End Time  \
0          7077.0 2018-11-01 00:04:00-05:00   
1          7333.0 2018-11-01 00:11:00-05:00   
2          7002.0 2018-11-01 00:13:00-05:00   
3          7140.0 2018-11-01 00:25:00-05:00   
4          7210.0 2018-11-01 00:20:00-05:00   

                           End Station Name  Bike Id      User Type  
0                        College Park South      NaN  Annual Member  
1                   King St E / Victoria St      NaN  Annual Member  
2                St. George St / Bloor St W      NaN  Annual Member  
3               Macpherson Ave / Spadina Rd      NaN  Annual Member  
4  Mary McCormick Rec Centre (Sheridan Ave)      NaN  Annual Member

Repeating the process for the `End Station Id`.

In [73]:
trips = trips.merge(missing_station, left_on=['End Station Name'], right_on = ['Data Station Name'], how = 'left')
trips['End Station Id'] = trips['End Station Id'].where(trips['End Station Id'].notnull(), trips['Station Id'])
trips = trips[['Trip Id', 'Subscription Id', 'Trip Duration', 'Start Station Id',
       'Start Time', 'Start Station Name', 'End Station Id', 'End Time',
       'End Station Name', 'Bike Id', 'User Type']]
trips.head()

Trip Id  Subscription Id  Trip Duration  Start Station Id  \
0  4357633              NaN            167            7089.0   
1  4357634              NaN            461            7260.0   
2  4357635              NaN            604            7281.0   
3  4357637              NaN           1133            7047.0   
4  4357638              NaN            753            7032.0   

                 Start Time                 Start Station Name  \
0 2018-11-01 00:01:00-05:00               Church St  / Wood St   
1 2018-11-01 00:03:00-05:00        Spadina Ave / Adelaide St W   
2 2018-11-01 00:03:00-05:00  Charles St W / Balmuto St - SMART   
3 2018-11-01 00:06:00-05:00      University Ave / Gerrard St W   
4 2018-11-01 00:07:00-05:00          Augusta Ave / Dundas St W   

   End Station Id                  End Time  \
0          7077.0 2018-11-01 00:04:00-05:00   
1          7333.0 2018-11-01 00:11:00-05:00   
2          7002.0 2018-11-01 00:13:00-05:00   
3          7140.0 2018-11-01 00:25:00-05:00   
4          7210.0 2018-11-01 00:20:00-05:00   

                           End Station Name  Bike Id      User Type  
0                        College Park South      NaN  Annual Member  
1                   King St E / Victoria St      NaN  Annual Member  
2                St. George St / Bloor St W      NaN  Annual Member  
3               Macpherson Ave / Spadina Rd      NaN  Annual Member  
4  Mary McCormick Rec Centre (Sheridan Ave)      NaN  Annual Member

In [74]:
trips = trips.dropna(axis=0, subset=['End Station Id'], how = 'all')  # dropping the station name that couldn't be matched

In [75]:
trips = trips[['Trip Id', 'Subscription Id', 'Trip Duration', 'Start Station Id',
       'Start Time', 'End Station Id', 'End Time',
        'Bike Id', 'User Type']]

Now lets remerge the station names to our list of station names, and use the "official" ones provided by the API.

In [76]:
trips = trips.merge(stations[['Station Name', 'Station Id']], left_on = ['Start Station Id'], right_on = ['Station Id'])

trips = trips.rename(columns = {'Station Name': 'Start Station Name'})

In [77]:
trips = trips.merge(stations[['Station Name', 'Station Id']], left_on = ['End Station Id'], right_on = ['Station Id'])

trips = trips.rename(columns = {'Station Name': 'End Station Name'})

In [78]:
trips = trips[['Trip Id', 'Subscription Id', 'Trip Duration', 'Start Station Id',
       'Start Time', 'Start Station Name', 'End Station Id', 'End Time',
       'End Station Name', 'Bike Id', 'User Type']]

trips.head()

Trip Id  Subscription Id  Trip Duration  Start Station Id  \
0  4357633              NaN            167            7089.0   
1  4357747              NaN           1457            7089.0   
2  4471981              NaN            247            7089.0   
3  1525862              NaN            696            7089.0   
4  1525865              NaN            676            7089.0   

                 Start Time    Start Station Name  End Station Id  \
0 2018-11-01 00:01:00-05:00  Church St  / Wood St          7077.0   
1 2018-11-01 03:43:00-05:00  Church St  / Wood St          7077.0   
2 2018-11-28 18:50:00-05:00  Church St  / Wood St          7077.0   
3 2017-08-06 20:25:00-05:00  Church St  / Wood St          7077.0   
4 2017-08-06 20:25:00-05:00  Church St  / Wood St          7077.0   

                   End Time    End Station Name  Bike Id      User Type  
0 2018-11-01 00:04:00-05:00  College Park South      NaN  Annual Member  
1 2018-11-01 04:07:00-05:00  College Park South      NaN  Annual Member  
2 2018-11-28 18:55:00-05:00  College Park South      NaN  Annual Member  
3 2017-08-06 20:36:00-05:00  College Park South      NaN         Casual  
4 2017-08-06 20:36:00-05:00  College Park South      NaN         Casual

### Removing Outliers

We'll use the interquartile range method to filter out outliers. First, lets remove trips less than.

In [79]:
trips = trips[trips['Trip Duration'] >60]

In [80]:
iqr = float(trips['Trip Duration'].quantile([0.75])) - float(trips['Trip Duration'].quantile([0.25]))

In [81]:
iqr

695.0

In [82]:
upper = float(trips['Trip Duration'].quantile([0.75])) + 1.5* iqr
lower = float(trips['Trip Duration'].quantile([0.25])) - 1.5* iqr

In [83]:
upper, lower

(2182.5, -597.5)

In [84]:
trips = trips[trips['Trip Duration'].between(lower, upper)]

### Fixing the User Type

Some columns don't have a valid user type.

In [85]:
trips['User Type'].unique()

array(['Annual Member', 'Casual', 'Member', 'Casual Member'], dtype=object)

In [86]:
trips[trips['User Type'] == 'Casual']

Trip Id  Subscription Id  Trip Duration  Start Station Id  \
3        1525862              NaN            696            7089.0   
4        1525865              NaN            676            7089.0   
5        1529528              NaN            252            7089.0   
72       1127716              NaN            280            7089.0   
75       1178610              NaN            198            7089.0   
...          ...              ...            ...               ...   
8401157   815726              NaN            766            7194.0   
8401158   816014              NaN            765            7194.0   
8401169  1124587              NaN           1420            7095.0   
8401170  1124589              NaN           1385            7095.0   
8402087  1065938              NaN            438            7084.0   

                       Start Time           Start Station Name  \
3       2017-08-06 20:25:00-05:00         Church St  / Wood St   
4       2017-08-06 20:25:00-05:00         Church St  / Wood St   
5       2017-08-07 16:27:00-05:00         Church St  / Wood St   
72      2017-06-09 14:03:00-05:00         Church St  / Wood St   
75      2017-06-17 21:50:00-05:00         Church St  / Wood St   
...                           ...                          ...   
8401157 2017-03-03 11:03:00-05:00   Mortimer Ave / Coxwell Ave   
8401158 2017-03-03 14:15:00-05:00   Mortimer Ave / Coxwell Ave   
8401169 2017-06-08 21:57:00-05:00  Danforth Ave / Ellerbeck St   
8401170 2017-06-08 21:58:00-05:00  Danforth Ave / Ellerbeck St   
8402087 2017-05-29 08:04:00-05:00   High Park - Grenadier Cafe   

         End Station Id                  End Time            End Station Name  \
3                7077.0 2017-08-06 20:36:00-05:00          College Park South   
4                7077.0 2017-08-06 20:36:00-05:00          College Park South   
5                7077.0 2017-08-07 16:32:00-05:00          College Park South   
72               7077.0 2017-06-09 14:07:00-05:00          College Park South   
75               7077.0 2017-06-17 21:53:00-05:00          College Park South   
...                 ...                       ...                         ...   
8401157          7084.0 2017-03-03 11:16:00-05:00  High Park - Grenadier Cafe   
8401158          7084.0 2017-03-03 14:27:00-05:00  High Park - Grenadier Cafe   
8401169          7084.0 2017-06-08 22:21:00-05:00  High Park - Grenadier Cafe   
8401170          7084.0 2017-06-08 22:21:00-05:00  High Park - Grenadier Cafe   
8402087          7084.0 2017-05-29 08:11:00-05:00  High Park - Grenadier Cafe   

         Bike Id User Type  
3            NaN    Casual  
4            NaN    Casual  
5            NaN    Casual  
72           NaN    Casual  
75           NaN    Casual  
...          ...       ...  
8401157      NaN    Casual  
8401158      NaN    Casual  
8401169      NaN    Casual  
8401170      NaN    Casual  
8402087      NaN    Casual  

[249394 rows x 11 columns]

This typo isn't a big deal since we can safely guess at the actual user type, however the next error represents ~ 12.5% of the data.

In [87]:
trips[trips['User Type'] == 'Member']

Trip Id  Subscription Id  Trip Duration  Start Station Id  \
6        1546937              NaN            386            7089.0   
7        1612225              NaN            205            7089.0   
8        1642131              NaN            175            7089.0   
9        1681438              NaN            210            7089.0   
10       1698158              NaN            280            7089.0   
...          ...              ...            ...               ...   
8402083  1214326              NaN            132            7084.0   
8403968   936366              NaN            278            7086.0   
8403969   999648              NaN            322            7086.0   
8403970   816666              NaN            269            7086.0   
8403971   836623              NaN            229            7086.0   

                       Start Time          Start Station Name  End Station Id  \
6       2017-08-09 18:43:00-05:00        Church St  / Wood St          7077.0   
7       2017-08-18 00:45:00-05:00        Church St  / Wood St          7077.0   
8       2017-08-21 17:39:00-05:00        Church St  / Wood St          7077.0   
9       2017-08-26 12:41:00-05:00        Church St  / Wood St          7077.0   
10      2017-08-28 15:32:00-05:00        Church St  / Wood St          7077.0   
...                           ...                         ...             ...   
8402083 2017-06-24 11:18:00-05:00  High Park - Grenadier Cafe          7084.0   
8403968 2017-04-24 08:47:00-05:00         High Park - West Rd          7084.0   
8403969 2017-05-13 16:05:00-05:00         High Park - West Rd          7084.0   
8403970 2017-03-03 18:42:00-05:00         High Park - West Rd          7084.0   
8403971 2017-03-16 14:14:00-05:00         High Park - West Rd          7084.0   

                         End Time            End Station Name  Bike Id  \
6       2017-08-09 18:50:00-05:00          College Park South      NaN   
7       2017-08-18 00:49:00-05:00          College Park South      NaN   
8       2017-08-21 17:42:00-05:00          College Park South      NaN   
9       2017-08-26 12:44:00-05:00          College Park South      NaN   
10      2017-08-28 15:36:00-05:00          College Park South      NaN   
...                           ...                         ...      ...   
8402083 2017-06-24 11:21:00-05:00  High Park - Grenadier Cafe      NaN   
8403968 2017-04-24 08:51:00-05:00  High Park - Grenadier Cafe      NaN   
8403969 2017-05-13 16:11:00-05:00  High Park - Grenadier Cafe      NaN   
8403970 2017-03-03 18:47:00-05:00  High Park - Grenadier Cafe      NaN   
8403971 2017-03-16 14:18:00-05:00  High Park - Grenadier Cafe      NaN   

        User Type  
6          Member  
7          Member  
8          Member  
9          Member  
10         Member  
...           ...  
8402083    Member  
8403968    Member  
8403969    Member  
8403970    Member  
8403971    Member  

[1142695 rows x 11 columns]

We can't safely guess at the user type, so we'll need to find a way to mitigate the effect of this error.

In [88]:
trips[trips['User Type'] == 'Casual'].max()

Trip Id                                 2383643
Subscription Id                             NaN
Trip Duration                              2182
Start Station Id                         7667.0
Start Time            2017-12-31 23:47:00-05:00
Start Station Name      York St / Queens Quay W
End Station Id                           7667.0
End Time              2018-01-01 00:11:00-05:00
End Station Name        York St / Queens Quay W
Bike Id                                     NaN
User Type                                Casual
dtype: object

In [89]:
trips[trips['User Type'] == 'Member'].max()

Trip Id                                 2383643
Subscription Id                             NaN
Trip Duration                              2182
Start Station Id                         7667.0
Start Time            2017-12-31 23:47:00-05:00
Start Station Name      York St / Queens Quay W
End Station Id                           7667.0
End Time              2018-01-01 00:11:00-05:00
End Station Name        York St / Queens Quay W
Bike Id                                     NaN
User Type                                Casual
dtype: object

Based on this, we can do analysis on the user type, but only for data after 2018. The errors are only present for trips before 2018. Lets set the User Type to None if its before 2018.

In [92]:
trips.loc[(trips['Start Time'] < pd.to_datetime('2018-01-01').tz_localize('EST')),'User Type'] = None

### Validating the Times

As a final check, lets see if the `Start Time` increases monotonically with `Trip Id`.

In [102]:
trips = trips.sort_values(by = 'Trip Id')

In [116]:
trips = trips.reset_index().drop(columns = ['index'])

In [117]:
trips['Start Time'].is_monotonic

False

Some rows aren't compliant, so lets find out which specific rows are causing this problem.

In [121]:
trips[trips['Start Time'].diff() < pd.Timedelta("0 days")]

Trip Id  Subscription Id  Trip Duration  Start Station Id  \
3968       717215              NaN           1465            7073.0   
4743       718215              NaN           1426            7029.0   
5277       718889              NaN            259            7028.0   
6168       719977              NaN            434            7029.0   
6187       720005              NaN            380            7008.0   
...           ...              ...            ...               ...   
8037535  10291994         812333.0            562            7028.0   
8037540  10291999         552193.0            402            7410.0   
8038641  10293299         553106.0           1042            7272.0   
8039043  10293763         863408.0           1101            7077.0   
8039049  10293769         863410.0            627            7202.0   

                       Start Time                Start Station Name  \
3968    2017-01-05 10:34:00-05:00      Spadina Ave / Fort York Blvd   
4743    2017-01-05 20:33:00-05:00   Bay St / Bloor St W (East Side)   
5277    2017-01-06 13:01:00-05:00              Gould St / Mutual St   
6168    2017-01-06 17:04:00-05:00   Bay St / Bloor St W (East Side)   
6187    2017-01-07 12:53:00-05:00  Wellesley St / Queen's Park Cres   
...                           ...                               ...   
8037535 2020-10-31 18:21:00-05:00              Gould St / Mutual St   
8037540 2020-10-31 18:21:00-05:00      Howard St / Rose Ave - SMART   
8038641 2020-10-31 21:02:00-05:00   Yonge St / Dundonald St - SMART   
8039043 2020-10-31 23:12:00-05:00                College Park South   
8039049 2020-10-31 23:13:00-05:00  Queen St W / York St (City Hall)   

         End Station Id                  End Time  \
3968             7126.0 2017-01-05 10:58:00-05:00   
4743             7046.0 2017-01-05 20:57:00-05:00   
5277             7041.0 2017-01-06 13:06:00-05:00   
6168             7038.0 2017-01-06 17:11:00-05:00   
6187             7041.0 2017-01-07 13:00:00-05:00   
...                 ...                       ...   
8037535          7257.0 2020-10-31 18:31:00-05:00   
8037540          7051.0 2020-10-31 18:28:00-05:00   
8038641          7027.0 2020-10-31 21:19:00-05:00   
8039043          7269.0 2020-10-31 23:31:00-05:00   
8039049          7038.0 2020-10-31 23:23:00-05:00   

                        End Station Name  Bike Id      User Type  
3968            Yonge St / Yorkville Ave      NaN           None  
4743          Niagara St / Richmond St W      NaN           None  
5277                Edward St / Yonge St      NaN           None  
6168              Dundas St W / Yonge St      NaN           None  
6187                Edward St / Yonge St      NaN           None  
...                                  ...      ...            ...  
8037535     Dundas St W / St. Patrick St   4140.0  Annual Member  
8037540            Mutual St / Shuter St    462.0  Annual Member  
8038641       Beverley  St / Dundas St W    679.0  Annual Member  
8039043  Toronto Eaton Centre (Yonge St)    884.0  Casual Member  
8039049           Dundas St W / Yonge St   2706.0  Casual Member  

[18314 rows x 11 columns]

It appears that the `Trip Id` field doesn't monotonically increase with `Start Time`. Below is one example. 

There isn't documentation about why this occurs, and if this is an issue we should be concerned about, so for now, this quirk will remain in the dataset.

In [127]:
trips.loc[3967:3969,:]

Trip Id  Subscription Id  Trip Duration  Start Station Id  \
3967   717214              NaN            224            7053.0   
3968   717215              NaN           1465            7073.0   
3969   717217              NaN            250            7024.0   

                    Start Time            Start Station Name  End Station Id  \
3967 2017-01-05 10:35:00-05:00              Metro Hall Plaza          7015.0   
3968 2017-01-05 10:34:00-05:00  Spadina Ave / Fort York Blvd          7126.0   
3969 2017-01-05 10:36:00-05:00      Dundonald St / Church St          7030.0   

                      End Time                End Station Name  Bike Id  \
3967 2017-01-05 10:39:00-05:00  King St W / Bay St (West Side)      NaN   
3968 2017-01-05 10:58:00-05:00        Yonge St / Yorkville Ave      NaN   
3969 2017-01-05 10:40:00-05:00         Bay St / Wellesley St W      NaN   

     User Type  
3967      None  
3968      None  
3969      None

In [130]:
trips.loc[4742:4744,:]

Trip Id  Subscription Id  Trip Duration  Start Station Id  \
4742   718214              NaN            847            7009.0   
4743   718215              NaN           1426            7029.0   
4744   718216              NaN            212            7167.0   

                    Start Time                  Start Station Name  \
4742 2017-01-05 20:55:00-05:00               King St E / Jarvis St   
4743 2017-01-05 20:33:00-05:00     Bay St / Bloor St W (East Side)   
4744 2017-01-05 20:58:00-05:00  Queen St E / George St (Moss Park)   

      End Station Id                  End Time  \
4742          7031.0 2017-01-05 21:09:00-05:00   
4743          7046.0 2017-01-05 20:57:00-05:00   
4744          7109.0 2017-01-05 21:01:00-05:00   

                     End Station Name  Bike Id User Type  
4742          Jarvis St / Isabella St      NaN      None  
4743       Niagara St / Richmond St W      NaN      None  
4744  Seaton St / Dundas St E - SMART      NaN      None

# Reading Weather Data

Before we merge the weather data with the trips data, we need to read and clean the weather data.

In [94]:
weather_filenames = list(filter(lambda x: x. startswith('en_climate_hourly_ON_6158359'), os.listdir('project_data')))

print(weather_filenames[0:5])

['en_climate_hourly_ON_6158359_04-2017_P1H.csv', 'en_climate_hourly_ON_6158359_03-2017_P1H.csv', 'en_climate_hourly_ON_6158359_02-2020_P1H.csv', 'en_climate_hourly_ON_6158359_05-2020_P1H.csv', 'en_climate_hourly_ON_6158359_11-2017_P1H.csv']


In [95]:
weather = pd.concat([pd.read_csv('project_data/' + name) for name in weather_filenames])

In [96]:
weather.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 35064 entries, 0 to 743
Data columns (total 28 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Longitude (x)        35064 non-null  float64
 1   Latitude (y)         35064 non-null  float64
 2   Station Name         35064 non-null  object 
 3   Climate ID           35064 non-null  int64  
 4   Date/Time            35064 non-null  object 
 5   Year                 35064 non-null  int64  
 6   Month                35064 non-null  int64  
 7   Day                  35064 non-null  int64  
 8   Time                 35064 non-null  object 
 9   Temp (°C)            34605 non-null  float64
 10  Temp Flag            89 non-null     object 
 11  Dew Point Temp (°C)  34556 non-null  float64
 12  Dew Point Temp Flag  139 non-null    object 
 13  Rel Hum (%)          34564 non-null  float64
 14  Rel Hum Flag         131 non-null    object 
 15  Wind Dir (10s deg)   32679 non-null  f

Some fields, such as tempeature, contain null values. However, other columns may be useful so we'll keep the data as long as there's some useful data for each measurement. In addition, abscence of a field may indicate information. However, we will drop all the `Flag` fields, and location fields to make the data easier to work with.

The `Flag` field doesn't have documenation on what it means. It appears the only non-null value is `M`.

In [97]:
weather[~weather['Wind Dir Flag'].isna()]['Wind Dir Flag']

19     M
36     M
49     M
85     M
188    M
      ..
676    M
678    M
679    M
738    M
346    M
Name: Wind Dir Flag, Length: 1054, dtype: object

In [100]:
weather = weather[['Date/Time', 'Temp (°C)',
       'Dew Point Temp (°C)', 'Rel Hum (%)',
       'Wind Dir (10s deg)', 
       'Wind Spd (km/h)', 'Visibility (km)',
        'Stn Press (kPa)', 'Hmdx', 'Wind Chill', 'Weather']]

In [101]:
weather

Date/Time  Temp (°C)  Dew Point Temp (°C)  Rel Hum (%)  \
0    2017-04-01 00:00        3.0                  1.4         89.0   
1    2017-04-01 01:00        3.0                  1.2         88.0   
2    2017-04-01 02:00        2.8                  0.7         86.0   
3    2017-04-01 03:00        2.6                  0.5         86.0   
4    2017-04-01 04:00        2.8                  0.4         84.0   
..                ...        ...                  ...          ...   
739  2020-12-31 19:00        NaN                  NaN          NaN   
740  2020-12-31 20:00        NaN                  NaN          NaN   
741  2020-12-31 21:00        NaN                  NaN          NaN   
742  2020-12-31 22:00        NaN                  NaN          NaN   
743  2020-12-31 23:00        NaN                  NaN          NaN   

     Wind Dir (10s deg)  Wind Spd (km/h)  Visibility (km)  Stn Press (kPa)  \
0                   1.0             18.0             16.1           100.13   
1                  35.0             17.0             16.1           100.19   
2                  35.0             18.0             16.1           100.25   
3                  35.0             22.0             16.1           100.32   
4                  36.0             13.0             16.1           100.39   
..                  ...              ...              ...              ...   
739                 NaN              NaN              NaN              NaN   
740                 NaN              NaN              NaN              NaN   
741                 NaN              NaN              NaN              NaN   
742                 NaN              NaN              NaN              NaN   
743                 NaN              NaN              NaN              NaN   

     Hmdx  Wind Chill Weather  
0     NaN         NaN     NaN  
1     NaN         NaN     NaN  
2     NaN         NaN     NaN  
3     NaN         NaN     NaN  
4     NaN         NaN     NaN  
..    ...         ...     ...  
739   NaN         NaN     NaN  
740   NaN         NaN     NaN  
741   NaN         NaN     NaN  
742   NaN         NaN     NaN  
743   NaN         NaN     NaN  

[35064 rows x 11 columns]

The last step is to convert the `Date/Time` column to a `DatetimeIndex` with an EST timezone.

In [169]:
weather['Date/Time'] = pd.DatetimeIndex(weather['Date/Time']).tz_localize(tz = 'EST')

# Merging the Datasets

Now we'll need to merge the datasets. We'll have to first convert the timestamps for the trips to the nearest hour. This will be in a new column names `merge_time`.

In [165]:
trips['merge_time'] = trips['Start Time'].round('h')

Now we can merge the datasets.

In [172]:
trips = trips.merge(weather, left_on = ['merge_time'], right_on = ['Date/Time'], how = 'left')

There were some trips that occured on December 31 from 8PM to Midnight. This is likely due to the way Bikeshare Toronto saves the times and timezones. Lets remove those rows since we don't have weather data for those trips.

In [179]:
trips = trips[(trips['Start Time'] > pd.to_datetime('2017-01-01').tz_localize('EST'))]

In [180]:
trips

Trip Id  Subscription Id  Trip Duration  Start Station Id  \
58         712441              NaN            274            7006.0   
59         712442              NaN            538            7046.0   
60         712443              NaN            992            7048.0   
61         712444              NaN           1005            7177.0   
62         712445              NaN            645            7203.0   
...           ...              ...            ...               ...   
8039141  10293869         534405.0            497            7302.0   
8039142  10293870         630937.0            272            7129.0   
8039143  10293872         539030.0            646            7156.0   
8039144  10293874         863192.0            462            7026.0   
8039145  10293875         524592.0            726            7339.0   

                       Start Time  \
58      2017-01-01 00:03:00-05:00   
59      2017-01-01 00:03:00-05:00   
60      2017-01-01 00:05:00-05:00   
61      2017-01-01 00:09:00-05:00   
62      2017-01-01 00:14:00-05:00   
...                           ...   
8039141 2020-10-31 23:57:00-05:00   
8039142 2020-10-31 23:57:00-05:00   
8039143 2020-10-31 23:58:00-05:00   
8039144 2020-10-31 23:59:00-05:00   
8039145 2020-10-31 23:59:00-05:00   

                                    Start Station Name  End Station Id  \
58                     Bay St / College St (East Side)          7021.0   
59                          Niagara St / Richmond St W          7147.0   
60         Front St W / Yonge St (Hockey Hall of Fame)          7089.0   
61                     East Liberty St / Pirandello St          7202.0   
62       Bathurst St/Queens Quay(Billy Bishop Airport)          7010.0   
...                                                ...             ...   
8039141               Winona Dr / Davenport Rd - SMART          7144.0   
8039142                       Davenport Rd / Avenue Rd          7130.0   
8039143                 Salem Ave / Bloor St W - SMART          7232.0   
8039144                         Bay St / St. Joseph St          7453.0   
8039145                    Carlaw Ave / Strathcona Ave          7311.0   

                         End Time                      End Station Name  \
58      2017-01-01 00:08:00-05:00                    Bay St / Albert St   
59      2017-01-01 00:12:00-05:00                King St W / Fraser Ave   
60      2017-01-01 00:22:00-05:00                  Church St  / Wood St   
61      2017-01-01 00:26:00-05:00      Queen St W / York St (City Hall)   
62      2017-01-01 00:25:00-05:00               King St W / Spadina Ave   
...                           ...                                   ...   
8039141 2020-01-11 00:05:00-05:00  Palmerston Ave / Vermont Ave - SMART   
8039142 2020-01-11 00:02:00-05:00             Davenport Rd / Bedford Rd   
8039143 2020-01-11 00:09:00-05:00     Queen St W  /  Fuller Ave - SMART   
8039144 2020-01-11 00:06:00-05:00               Jarvis St / Maitland Pl   
8039145 2020-01-11 00:11:00-05:00           Sherbourne St / Isabella St   

         Bike Id  ... Temp (°C) Dew Point Temp (°C) Rel Hum (%)  \
58           NaN  ...       1.5                -3.6        69.0   
59           NaN  ...       1.5                -3.6        69.0   
60           NaN  ...       1.5                -3.6        69.0   
61           NaN  ...       1.5                -3.6        69.0   
62           NaN  ...       1.5                -3.6        69.0   
...          ...  ...       ...                 ...         ...   
8039141   6125.0  ...       8.9                 3.7        70.0   
8039142   3871.0  ...       8.9                 3.7        70.0   
8039143    304.0  ...       8.9                 3.7        70.0   
8039144   3459.0  ...       8.9                 3.7        70.0   
8039145   4057.0  ...       8.9                 3.7        70.0   

         Wind Dir (10s deg)  Wind Spd (km/h)  Visibility (km)  \
58                     26.0             39.0        

In [182]:
trips.to_csv('trips_raw_data.csv')

# Aggregating Data

For ease of use for the rest of my group, we're going to aggregate the data to the neares hour. 

In [193]:
hourly_trips = trips
hourly_trips['Start Time'] = hourly_trips['merge_time']
hourly_trips = hourly_trips[['Trip Id', 'Trip Duration', 'Start Station Id',
       'Start Time',  'End Station Id', 'End Time',
         'User Type']]

In [194]:
hourly_trips

Trip Id  Trip Duration  Start Station Id                Start Time  \
58         712441            274            7006.0 2017-01-01 00:00:00-05:00   
59         712442            538            7046.0 2017-01-01 00:00:00-05:00   
60         712443            992            7048.0 2017-01-01 00:00:00-05:00   
61         712444           1005            7177.0 2017-01-01 00:00:00-05:00   
62         712445            645            7203.0 2017-01-01 00:00:00-05:00   
...           ...            ...               ...                       ...   
8039141  10293869            497            7302.0 2020-11-01 00:00:00-05:00   
8039142  10293870            272            7129.0 2020-11-01 00:00:00-05:00   
8039143  10293872            646            7156.0 2020-11-01 00:00:00-05:00   
8039144  10293874            462            7026.0 2020-11-01 00:00:00-05:00   
8039145  10293875            726            7339.0 2020-11-01 00:00:00-05:00   

         End Station Id                  End Time      User Type  
58               7021.0 2017-01-01 00:08:00-05:00           None  
59               7147.0 2017-01-01 00:12:00-05:00           None  
60               7089.0 2017-01-01 00:22:00-05:00           None  
61               7202.0 2017-01-01 00:26:00-05:00           None  
62               7010.0 2017-01-01 00:25:00-05:00           None  
...                 ...                       ...            ...  
8039141          7144.0 2020-01-11 00:05:00-05:00  Annual Member  
8039142          7130.0 2020-01-11 00:02:00-05:00  Annual Member  
8039143          7232.0 2020-01-11 00:09:00-05:00  Annual Member  
8039144          7453.0 2020-01-11 00:06:00-05:00  Casual Member  
8039145          7311.0 2020-01-11 00:11:00-05:00  Annual Member  

[8039088 rows x 7 columns]

In [195]:
hourly_trips = hourly_trips.join(hourly_trips['User Type'].str.get_dummies())

In [196]:
hourly_trips = hourly_trips.drop(columns = ['User Type'])

In [198]:
hourly_trips

Trip Id  Trip Duration  Start Station Id                Start Time  \
58         712441            274            7006.0 2017-01-01 00:00:00-05:00   
59         712442            538            7046.0 2017-01-01 00:00:00-05:00   
60         712443            992            7048.0 2017-01-01 00:00:00-05:00   
61         712444           1005            7177.0 2017-01-01 00:00:00-05:00   
62         712445            645            7203.0 2017-01-01 00:00:00-05:00   
...           ...            ...               ...                       ...   
8039141  10293869            497            7302.0 2020-11-01 00:00:00-05:00   
8039142  10293870            272            7129.0 2020-11-01 00:00:00-05:00   
8039143  10293872            646            7156.0 2020-11-01 00:00:00-05:00   
8039144  10293874            462            7026.0 2020-11-01 00:00:00-05:00   
8039145  10293875            726            7339.0 2020-11-01 00:00:00-05:00   

         End Station Id                  End Time  Annual Member  \
58               7021.0 2017-01-01 00:08:00-05:00              0   
59               7147.0 2017-01-01 00:12:00-05:00              0   
60               7089.0 2017-01-01 00:22:00-05:00              0   
61               7202.0 2017-01-01 00:26:00-05:00              0   
62               7010.0 2017-01-01 00:25:00-05:00              0   
...                 ...                       ...            ...   
8039141          7144.0 2020-01-11 00:05:00-05:00              1   
8039142          7130.0 2020-01-11 00:02:00-05:00              1   
8039143          7232.0 2020-01-11 00:09:00-05:00              1   
8039144          7453.0 2020-01-11 00:06:00-05:00              0   
8039145          7311.0 2020-01-11 00:11:00-05:00              1   

         Casual Member  
58                   0  
59                   0  
60                   0  
61                   0  
62                   0  
...                ...  
8039141              0  
8039142              0  
8039143              0  
8039144              1  
8039145              0  

[8039088 rows x 8 columns]

In [201]:
hourly_trips.groupby(['Start Time', 'Start Station Id', 'End Station Id']).agg({'Trip Id':'count',
'Annual Member': 'sum','Casual Member':'sum', 'Trip Duration':'mean'})

Trip Id  \
Start Time                Start Station Id End Station Id            
2017-01-01 00:00:00-05:00 7000.0           7070.0                1   
                          7003.0           7003.0                1   
                          7006.0           7021.0                1   
                          7023.0           7102.0                2   
                          7046.0           7147.0                1   
...                                                            ...   
2020-12-11 00:00:00-05:00 7502.0           7078.0                1   
                          7526.0           7078.0                1   
                          7563.0           7561.0                2   
                          7569.0           7059.0                1   
                          7586.0           7589.0                1   

                                                           Annual Member  \
Start Time                Start Station Id End Station Id                  
2017-01-01 00:00:00-05:00 7000.0           7070.0                      0   
                          7003.0           7003.0                      0   
                          7006.0           7021.0                      0   
                          7023.0           7102.0                      0   
                          7046.0           7147.0                      0   
...                                                                  ...   
2020-12-11 00:00:00-05:00 7502.0           7078.0                      1   
                          7526.0           7078.0                      1   
                          7563.0           7561.0                      2   
                          7569.0           7059.0                      0   
                          7586.0           7589.0                      1   

                                                           Casual Member  \
Start Time                Start Station Id End Station Id                  
2017-01-01 00:00:00-05:00 7000.0           7070.0                      0   
                          7003.0           7003.0                      0   
                          7006.0           7021.0                      0   
                          7023.0           7102.0                      0   
                          7046.0           7147.0                      0   
...                                                                  ...   
2020-12-11 00:00:00-05:00 7502.0           7078.0                      0   
                          7526.0           7078.0                      0   
                          7563.0           7561.0                      0   
                          7569.0           7059.0                      1   
                          7586.0           7589.0                      0   

                                                           Trip Duration  
Start Time                Start Station Id End Station Id                 
2017-01-01 00:00:00-05:00 7000.0           7070.0                  348.0  
                          7003.0           7003.0                  625.0  
                          7006.0           7021.0                  274.0  
                          7023.0           7102.0                  613.0  
                          7046.0           7147.0                  538.0  
...                                                                  ...  
2020-12-11 00:00:00-05:00 7502.0           7078.0                  288.0  
                          7526.0           7078.0                  805.0  
                          7563.0           7561.0                  540.0  
                          7569.0           7059.0                  745.0  
                          7586.0           7589.0                 1078.0  

[7064321 rows x 4 columns]